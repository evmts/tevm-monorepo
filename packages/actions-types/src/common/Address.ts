import type { Address as _Address } from '@tevm/utils'

/**
 * An ethereum address represented as a hex string
 * @see https://abitype.dev/config#addresstype for configuration options to change type to being a string if preferred
 */
export type Address = _Address
