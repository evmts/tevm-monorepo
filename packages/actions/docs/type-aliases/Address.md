[**@tevm/actions**](../README.md)

***

[@tevm/actions](../globals.md) / Address

# Type Alias: Address

> **Address**: `_Address`

Defined in: [packages/actions/src/common/Address.ts:7](https://github.com/evmts/tevm-monorepo/blob/main/packages/actions/src/common/Address.ts#L7)

An ethereum address represented as a hex string

## See

https://abitype.dev/config#addresstype for configuration options to change type to being a string if preferred
