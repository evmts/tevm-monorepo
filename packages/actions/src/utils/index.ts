export * from './SerializeToJson.js'
export * from './blockToJsonRpcBlock.js'
export * from './generateRandomId.js'
export * from './parseBlockTag.js'
export * from './txToJSONRPCTx.js'
