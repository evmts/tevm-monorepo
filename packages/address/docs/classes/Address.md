[**@tevm/address**](../README.md) • **Docs**

***

[@tevm/address](../globals.md) / Address

# Class: Address

Utility class for ethereum addresses.
Wraps EthjsAddress with a tevm style API.
toString returns a checksummed address rather than lowercase

## Example

```javascript
import { createAddress } from '@tevm/address';

// takes hex string
let address = createAddress(`0x${'00'.repeat(20)}`);
// takes number and bigint
address = createAddress(0);
// takes bytes
address = createAddress(new Uint8Array());
// non hex string
address = createAddress('55'.repeat(20));
```

## Extends

- `Address`

## Constructors

### new Address()

> **new Address**(`bytes`): [`Address`](Address.md)

#### Parameters

• **bytes**: `Uint8Array`

#### Returns

[`Address`](Address.md)

#### Inherited from

`EthjsAddress.constructor`

#### Defined in

node\_modules/.pnpm/@ethereumjs+util@9.0.3/node\_modules/@ethereumjs/util/dist/esm/address.d.ts:6

## Properties

### bytes

> `readonly` **bytes**: `Uint8Array`

#### Inherited from

`EthjsAddress.bytes`

#### Defined in

node\_modules/.pnpm/@ethereumjs+util@9.0.3/node\_modules/@ethereumjs/util/dist/esm/address.d.ts:5

## Methods

### equals()

> **equals**(`address`): `boolean`

Is address equal to another.

#### Parameters

• **address**: `Address`

#### Returns

`boolean`

#### Inherited from

`EthjsAddress.equals`

#### Defined in

node\_modules/.pnpm/@ethereumjs+util@9.0.3/node\_modules/@ethereumjs/util/dist/esm/address.d.ts:42

***

### isPrecompileOrSystemAddress()

> **isPrecompileOrSystemAddress**(): `boolean`

True if address is in the address range defined
by EIP-1352

#### Returns

`boolean`

#### Inherited from

`EthjsAddress.isPrecompileOrSystemAddress`

#### Defined in

node\_modules/.pnpm/@ethereumjs+util@9.0.3/node\_modules/@ethereumjs/util/dist/esm/address.d.ts:51

***

### isZero()

> **isZero**(): `boolean`

Is address zero.

#### Returns

`boolean`

#### Inherited from

`EthjsAddress.isZero`

#### Defined in

node\_modules/.pnpm/@ethereumjs+util@9.0.3/node\_modules/@ethereumjs/util/dist/esm/address.d.ts:46

***

### toBytes()

> **toBytes**(): `Uint8Array`

Returns a new Uint8Array representation of address.

#### Returns

`Uint8Array`

#### Inherited from

`EthjsAddress.toBytes`

#### Defined in

node\_modules/.pnpm/@ethereumjs+util@9.0.3/node\_modules/@ethereumjs/util/dist/esm/address.d.ts:59

***

### toString()

> **toString**(): \`0x$\{string\}\`

Returns the checksummed address.

#### Returns

\`0x$\{string\}\`

The checksummed address.

#### Overrides

`EthjsAddress.toString`

#### Defined in

[packages/address/src/Address.js:27](https://github.com/evmts/tevm-monorepo/blob/main/packages/address/src/Address.js#L27)

***

### fromPrivateKey()

> `static` **fromPrivateKey**(`privateKey`): `Address`

Returns an address for a given private key.

#### Parameters

• **privateKey**: `Uint8Array`

A private key must be 256 bits wide

#### Returns

`Address`

#### Inherited from

`EthjsAddress.fromPrivateKey`

#### Defined in

node\_modules/.pnpm/@ethereumjs+util@9.0.3/node\_modules/@ethereumjs/util/dist/esm/address.d.ts:25

***

### fromPublicKey()

> `static` **fromPublicKey**(`pubKey`): `Address`

Returns an address for a given public key.

#### Parameters

• **pubKey**: `Uint8Array`

The two points of an uncompressed key

#### Returns

`Address`

#### Inherited from

`EthjsAddress.fromPublicKey`

#### Defined in

node\_modules/.pnpm/@ethereumjs+util@9.0.3/node\_modules/@ethereumjs/util/dist/esm/address.d.ts:20

***

### fromString()

> `static` **fromString**(`str`): `Address`

Returns an Address object from a hex-encoded string.

#### Parameters

• **str**: `string`

Hex-encoded address

#### Returns

`Address`

#### Inherited from

`EthjsAddress.fromString`

#### Defined in

node\_modules/.pnpm/@ethereumjs+util@9.0.3/node\_modules/@ethereumjs/util/dist/esm/address.d.ts:15

***

### generate()

> `static` **generate**(`from`, `nonce`): `Address`

Generates an address for a newly created contract.

#### Parameters

• **from**: `Address`

The address which is creating this new address

• **nonce**: `bigint`

The nonce of the from account

#### Returns

`Address`

#### Inherited from

`EthjsAddress.generate`

#### Defined in

node\_modules/.pnpm/@ethereumjs+util@9.0.3/node\_modules/@ethereumjs/util/dist/esm/address.d.ts:31

***

### generate2()

> `static` **generate2**(`from`, `salt`, `initCode`): `Address`

Generates an address for a contract created using CREATE2.

#### Parameters

• **from**: `Address`

The address which is creating this new address

• **salt**: `Uint8Array`

A salt

• **initCode**: `Uint8Array`

The init code of the contract being created

#### Returns

`Address`

#### Inherited from

`EthjsAddress.generate2`

#### Defined in

node\_modules/.pnpm/@ethereumjs+util@9.0.3/node\_modules/@ethereumjs/util/dist/esm/address.d.ts:38

***

### zero()

> `static` **zero**(): `Address`

Returns the zero address.

#### Returns

`Address`

#### Inherited from

`EthjsAddress.zero`

#### Defined in

node\_modules/.pnpm/@ethereumjs+util@9.0.3/node\_modules/@ethereumjs/util/dist/esm/address.d.ts:10
