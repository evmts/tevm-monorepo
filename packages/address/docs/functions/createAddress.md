[**@tevm/address**](../README.md) â€¢ **Docs**

***

[@tevm/address](../globals.md) / createAddress

# Function: createAddress()

> **createAddress**(`address`): [`Address`](../classes/Address.md)

Creates an [Address](../classes/Address.md) for safely interacting with an Ethereum
Wraps EthjsAddress with a tevm style API.
toString returns a checksummed address rather than lowercase

## Parameters

â€¢ **address**: `number` \| `bigint` \| `AddressLike`

## Returns

[`Address`](../classes/Address.md)

## Example

```typescript
import { createAddress } from '@tevm/address'`

// takes hex string
let address = createAddress(`0x${'00'.repeat(20)}`)
// takes number and bigint
address = createAddress(0)
// takes bytes
address = createAddress(new Uint8Array()))
// non hex string
address = createAddress('55'.repeat(20))
```

## Throws

if the input is not a valid address}

## Defined in

[packages/address/src/createAddress.js:27](https://github.com/evmts/tevm-monorepo/blob/main/packages/address/src/createAddress.js#L27)
