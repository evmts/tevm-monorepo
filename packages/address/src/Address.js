import { EthjsAddress, bytesToHex, getAddress } from '@tevm/utils'

/**
 * Utility class for ethereum addresses.
 * Wraps {@link EthjsAddress} with a tevm style API.
 * toString returns a checksummed address rather than lowercase
 * @example
 * ```javascript
 * import { createAddress } from '@tevm/address';
 *
 * // takes hex string
 * let address = createAddress(`0x${'00'.repeat(20)}`);
 * // takes number and bigint
 * address = createAddress(0);
 * // takes bytes
 * address = createAddress(new Uint8Array());
 * // non hex string
 * address = createAddress('55'.repeat(20));
 * ```
 */
export class Address extends EthjsAddress {
	/**
	 * Returns the checksummed address.
	 * @override
	 * @returns {import('@tevm/utils').Address} The checksummed address.
	 */
	toString() {
		return getAddress(bytesToHex(this.bytes))
	}
}
