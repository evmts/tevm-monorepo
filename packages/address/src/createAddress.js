import { InvalidAddressError, UnreachableCodeError } from '@tevm/errors'
import { EthjsAddress, hexToBytes } from '@tevm/utils'
import { numberToBytes } from 'viem'
import { Address } from './Address.js'

/**
 * Creates an {@link Address} for safely interacting with an Ethereum
 * Wraps {@link EthjsAddress} with a tevm style API.
 * toString returns a checksummed address rather than lowercase
 * @example
 * ```typescript
 * import { createAddress } from '@tevm/address'`
 *
 * // takes hex string
 * let address = createAddress(`0x${'00'.repeat(20)}`)
 * // takes number and bigint
 * address = createAddress(0)
 * // takes bytes
 * address = createAddress(new Uint8Array()))
 * // non hex string
 * address = createAddress('55'.repeat(20))
 * ```
 * @param {import("@tevm/utils").AddressLike | number | bigint | string} address
 * @returns {import('./Address.js').Address}
 * @throws {InvalidAddressError} if the input is not a valid address}
 */
export const createAddress = (address) => {
	try {
		if (address instanceof EthjsAddress) {
			return new Address(address.bytes)
		}
		if (address instanceof Uint8Array) {
			return new Address(address)
		}
		if (typeof address === 'number' || typeof address === 'bigint') {
			return new Address(numberToBytes(address, { size: 20 }))
		}
		if (typeof address === 'string' && address.startsWith('0x')) {
			return new Address(hexToBytes(/** @type {import('viem').Hex}*/ (address), { size: 20 }))
		}
		if (typeof address === 'string') {
			return new Address(hexToBytes(`0x${address}`, { size: 20 }))
		}
		throw new UnreachableCodeError(address, `Receieved an unexpected input for createAddress ${address}`)
	} catch (e) {
		if (e instanceof UnreachableCodeError) {
			throw new InvalidAddressError(
				'Recieved an invalid address input type for createAddress. Valid input types include hex string, unprefixed hex, bytes, number, bigint, or EthjsAddress',
				{ cause: e },
			)
		}
		if (e instanceof Error) {
			throw new InvalidAddressError(`Recieved an invalid address input ${e.message}`, { cause: e })
		}
		throw new InvalidAddressError(`Recieved an invalid address input ${address}`, { cause: /** @type {any} */ (e) })
	}
}
