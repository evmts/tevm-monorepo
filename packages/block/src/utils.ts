import { EthjsAddress, keccak256, toType as originalToType, toBytes } from '@tevm/utils'

/**
 * Creates a zero-filled buffer of the specified length
 * @param {number} length - The length of the buffer
 * @returns {Uint8Array} A zero-filled Uint8Array
 */
export function zeros(length: number): Uint8Array {
	return new Uint8Array(length)
}

/**
 * Creates the zero address
 * @returns {EthjsAddress} The zero address
 */
export function createZeroAddress() {
	return new EthjsAddress(new Uint8Array(20))
}

/**
 * Creates an address from a public key
 * @param {Uint8Array} publicKey - The public key
 * @returns {EthjsAddress} The address
 */
export function createAddressFromPublicKey(publicKey: Uint8Array): EthjsAddress {
	// Create address from public key by taking the last 20 bytes of keccak256 hash
	const addressBytes = keccak256(publicKey, 'bytes').slice(-20)
	return new EthjsAddress(addressBytes)
}

/**
 * Wrapper for toType that handles string inputs properly
 * @param {any} input - The input value
 * @param {number} outputType - The output type
 * @returns {any} The converted value
 */
export function safeToType(input: any, outputType: number): any {
	if (input === null || input === undefined) {
		return undefined
	}

	// Handle string inputs by converting to bytes first
	if (typeof input === 'string' && outputType === 2) {
		const bytes = toBytes(input)
		return originalToType(bytes, outputType as any)
	}

	return originalToType(input, outputType as any)
}

/**
 * Gets the v value from a signature
 * @param {any} signature - The signature object
 * @returns {bigint} The v value
 */
export function getSignatureV(signature: any): bigint {
	// Handle both old format (v) and new format (recovery)
	if (signature.v !== undefined) {
		return BigInt(signature.v)
	}
	if (signature.recovery !== undefined) {
		return BigInt(signature.recovery) + 27n
	}
	throw new Error('Invalid signature format: missing v or recovery')
}
