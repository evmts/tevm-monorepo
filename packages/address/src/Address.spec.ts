import { EthjsAddress, getAddress, hexToBytes } from '@tevm/utils'
import { describe, expect, it } from 'vitest'
import { Address } from './Address.js'
import { createAddress } from './createAddress.js'

describe('Address', () => {
	it('should create an address from 0 string', () => {
		const address = createAddress(0)
		expect(address).toBeInstanceOf(Address)
		expect(address).toBeInstanceOf(EthjsAddress)
		expect(address.toString()).toMatchSnapshot()
		expect(address.bytes).toMatchSnapshot()
		expect(address.equals(address)).toBe(true)
		expect(address.equals(createAddress(25))).toBe(false)
		expect(address.isPrecompileOrSystemAddress()).toBe(true)
		expect(address.isZero()).toBe(true)
	})

	it('should create an address from non zero', () => {
		const address = createAddress(98765)
		expect(address).toBeInstanceOf(Address)
		expect(address).toBeInstanceOf(EthjsAddress)
		expect(address).toBeInstanceOf(EthjsAddress)
		expect(address.toString()).toMatchSnapshot()
		expect(address.bytes).toMatchSnapshot()
		expect(address.equals(address)).toBe(true)
		expect(address.isPrecompileOrSystemAddress()).toBe(false)
		expect(address.isZero()).toBe(false)
	})

	it('should return checksummed address as a string', () => {
		const hexAddress = `0x${'a'.repeat(40)}` as const
		const address = new Address(hexToBytes(hexAddress))
		expect(address.toString()).not.toBe(hexAddress)
		expect(address.toString()).toBe(getAddress(hexAddress))
	})
})
