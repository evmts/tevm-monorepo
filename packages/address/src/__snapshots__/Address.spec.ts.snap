// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Address should create an address from 0 string 1`] = `"0x0000000000000000000000000000000000000000"`;

exports[`Address should create an address from 0 string 2`] = `
Uint8Array [
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
]
`;

exports[`Address should create an address from non zero 1`] = `"0x00000000000000000000000000000000000181cd"`;

exports[`Address should create an address from non zero 2`] = `
Uint8Array [
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  1,
  129,
  205,
]
`;
